### PR TITLE
NOISSUE - Add pure JSON transformer 

### PR DESCRIPTION
**Pull request by manuio**
Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

* JSON key-value pairs of the message are transformed as a single Mainflux message where "key" and "value" will be represented as "name" and "Value, StringValue, DataValue or BoolValue". key-value pairs can be also filtered. 

* For example, this JSON message:
```
{"device":"1234", "type": "temperature", "data": {"value": 1, "ts": 12345678}}
``` 

will be converted in 4 transformers.Message:
```
{
  "channel": "9a077190-0853-4d64-85f2-d5efc6cdf92d",
  "publisher": "f3845ea2-5a3f-40e5-8fa3-9cdef6ac4f30",
  "protocol": "http",
  "name": "type",
  "time": 1602499859.975915,
  "string_value": "temperature"
}

{
  "channel": "9a077190-0853-4d64-85f2-d5efc6cdf92d",
  "publisher": "f3845ea2-5a3f-40e5-8fa3-9cdef6ac4f30",
  "protocol": "http",
  "name": "device",
  "time": 1602499859.975915,
  "string_value": "1234"
}

{
  "channel": "9a077190-0853-4d64-85f2-d5efc6cdf92d",
  "publisher": "f3845ea2-5a3f-40e5-8fa3-9cdef6ac4f30",
  "protocol": "http",
  "name": "data/value",
  "time": 1602499859.975915,
  "value": 1
}

{
  "channel": "9a077190-0853-4d64-85f2-d5efc6cdf92d",
  "publisher": "f3845ea2-5a3f-40e5-8fa3-9cdef6ac4f30",
  "protocol": "http",
  "name": "data/ts",
  "time": 1602499859.975915,
  "value": 12345678
}
```

### CODE REVIEW

File: docker/addons/cassandra-writer/filters.toml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# To listen all messsage broker subjects use default value "channels.>".
+# For specific filtering pass a list of subjects starting by "channels." and
+# followed by a subtopic (e.g ["channels.<channel_id>.sub.topic.x", ...]).
+[subjects]
+filter = ["channels.>"]
+
+# If writer is used with ContentType application/json JSON key-value pairs can
+# be filtered. Keys are used to store correspondinng pair. (e.g ["temperature", "humidity"]).
```
Comment (mteodor) [inline docker/addons/cassandra-writer/filters.toml]: this means that only the enlisted keys will be saved, right?

File: pkg/transformers/json/transformer.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+	err := json.Unmarshal(mm.Payload, &payload)
+	if err != nil {
+		return []transformers.Message{}, errors.Wrap(errDecode, err)
+	}
+
+	var msgs []transformers.Message
+
+	for k, v := range payload {
+		// Apply key filter
+		for _, tk := range t.keys {
```
Comment (mteodor) [inline pkg/transformers/json/transformer.go]: how about using a map for searching?

File: pkg/transformers/json/transformer.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+	return msgs, nil
+}
+
+func createKeyValMsg(mm messaging.Message, key string, val interface{}) []transformers.Message {
+	msg := transformers.Message{
+		Channel:   mm.Channel,
+		Subtopic:  mm.Subtopic,
+		Publisher: mm.Publisher,
+		Protocol:  mm.Protocol,
+		Name:      key,
```
Comment (mteodor) [inline pkg/transformers/json/transformer.go]: is there a way to group records by incoming message, as I understand incoming JSON payload that has like 10 fields will produce 10 records in writers database, but do we have a way to group these 10 records?

File: docker/addons/cassandra-writer/filters.toml (hunk 1 above)
Comment (manuio) [reply docker/addons/cassandra-writer/filters.toml]: Right! Or all if listed key is "*"

**Comment by drasko** [comment]
Can you please write some examples in the PR description, so that we can understand better these mappings and what is produced?

File: pkg/transformers/message.go [hunk 4]
```diff
@@ -1,4 +1,4 @@
-package senml
+package transformers
```
Comment (drasko) [inline pkg/transformers/message.go]: I am not sure that `Message{}` should be in `transformers` package.

Maybe it should not be in `senml` neither, but I am really skeptical about this one.

**Review by drasko — CHANGES_REQUESTED** [review]

File: docker/addons/cassandra-writer/filters.toml [hunk 5]
```diff
@@ -0,0 +1,12 @@
+# To listen all messsage broker subjects use default value "channels.>".
```
[the remark this answers is not in the document]
Comment (manuio) [reply docker/addons/cassandra-writer/filters.toml]: I agree

File: pkg/transformers/message.go (hunk 4 above)
Comment (manuio) [reply pkg/transformers/message.go]: Let's think about. This is more a DB message format.

File: pkg/transformers/json/transformer.go (hunk 3 above)
Comment (mteodor) [reply pkg/transformers/json/transformer.go]: what I meant is that you can easily determine that group of 10 records is from one `JSON` message.

Comment (manuio) [reply pkg/transformers/json/transformer.go]: Yes, for example with a reader JSON transformer that will group messages by timestamp (kind of state) and re-create the corresponding JSON

**Comment by manuio** [comment]
Closing this PR.  @dusanb94 will continue JSON transformer in a separate branch and we will merge later.

**Comment by drasko** [comment]
Actually, we can keep this PR open as it is opened from the `json` branch. @dusanb94 will continue working on this branch, improving the current PR.

**Comment by drasko** [comment]
Closed in favor of https://github.com/mainflux/mainflux/pull/1260

**Pull request closed without merging**